src/coreComponents/linearAlgebra/utilities/AsyncRequest.hpp

```diff
@@ -0,0 +1,70 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+/**
+ * @file AsyncRequest.hpp
+ */
+
+#ifndef GEOSX_LINEARALGEBRA_UTILITIES_ASYNCREQUEST_HPP_
+#define GEOSX_LINEARALGEBRA_UTILITIES_ASYNCREQUEST_HPP_
+
+#include "common/MpiWrapper.hpp"
+
+namespace geosx
+{
+
+/**
+ * @brief Helper class for managing non-blocking operation.
+ * @tparam T the type of the result of the non-blocking operation
+ */
+template< typename T >
+class AsyncRequest
+{
+public:
+
+  /**
+   * @brief Constructor.
+   * @tparam F the type of the lambda function
+   * @param f lambda function that defines the MPI non-blocking operation to be performed
+   */
+  template< typename F >
+  explicit AsyncRequest( F f )
+    : m_function( std::move( f ) )
+  {
+    m_result = std::make_unique< T >( T{} );
+    m_request = std::make_unique< MPI_Request >();
+    m_function( *m_request, *m_result );
+  }
+
+  /**
+   * @brief Completes the managed MPI request.
+   * @return the result of the non blocking operation
+   */
+  T const & complete() const
```

Comment:
This does not sound like the correct wording.
`complete` usually means "_If not already completed, sets the value returned by `get`and related methods to the given value._" see java doc [here](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#complete-T-).

I would change this to `get` like https://en.cppreference.com/w/cpp/thread/future/get